### PR TITLE
Fill: allow for single player fill restrictive placement and sweeping

### PR DESCRIPTION
**Pull request by alwaysintreble**
## What is this fixing or adding?
title

## How was this tested?
generated the same seed for 300 players under 4 separate conditions. existing emerald code with and without using the all_state cache, and then the same after merging in the single player optimization.
[emerald cache single player opt.txt](https://github.com/ArchipelagoMW/Archipelago/files/13239743/emerald.cache.single.player.opt.txt)
[emerald cache.txt](https://github.com/ArchipelagoMW/Archipelago/files/13239744/emerald.cache.txt)
[emerald no cache.txt](https://github.com/ArchipelagoMW/Archipelago/files/13239745/emerald.no.cache.txt)
[emerald no cache single player opt.txt](https://github.com/ArchipelagoMW/Archipelago/files/13239746/emerald.no.cache.single.player.opt.txt)

### CODE REVIEW

**Comment by espeon65536** [comment]
Tested on main's OoT, observed ~20% speedup on 10 worlds default settings and ~35% speedup on 20 worlds, presumably would get even better with more worlds but I don't have infinite processing power to test it.

File: BaseClasses.py [hunk 1]
```diff
@@ -693,9 +693,10 @@ def can_reach(self,
                 spot = self.multiworld.get_region(spot, player)
         return spot.can_reach(self)
 
-    def sweep_for_events(self, key_only: bool = False, locations: Optional[Iterable[Location]] = None) -> None:
+    def sweep_for_events(self, key_only: bool = False, locations: Optional[Iterable[Location]] = None,
+                         player: Optional[int] = None) -> None:
```
Comment (Berserker66) [inline BaseClasses.py:697]: this seems like a doubling of API responsibilities.  Surely the same result can be achieved by just passing in a single player's locations?

Comment (alwaysintreble) [reply BaseClasses.py:697]: that's what it does already. i suppose i could remove that arg and have the method do a check that all locations belong to only a single player, but that would be slower so i'm not sure if it's worth it?

Comment (alwaysintreble) [reply BaseClasses.py]: reread it and understand what you meant now

**Review by Zunawe — APPROVED** [review]
Changes look about like what I did a while ago when the conversation that sparked this PR was had. I don't think my approval can speak for whether these changes forget behaviors that I don't know about, but I can say that it works very well for Emerald's purposes. Emerald is starting to hurt quite badly in `pre_fill` due to having to collect and sweep from every world in order to place a handful of its own items. Aside from creating a patch file, this is by far the slowest aspect of Emerald right now.

Shuffling both badges and HMs (16 items among 16 locations, split into two groups) on my current branch for Emerald takes

Without this PR:
- ~1.1 seconds per `pre_fill` in a 10-player game
- ~10 minutes per `pre_fill` in a 2000-player game

With this PR
- ~0.3 seconds per `pre_fill` in a 10-player game
- ~2 minutes per `pre_fill` in a 2000-player game with this PR merged

It's another good chunk of time knocked off if there's an api to create single-player `all_state`s, though not nearly as much.

**Comment by alwaysintreble** [comment]
> It's another good chunk of time knocked off if there's an api to create single-player `all_state`s, though not nearly as much.

i have this https://github.com/ArchipelagoMW/Archipelago/pull/2427 but i'm not happy with it atm so would definitely like some feedback on it on how it can be improved/made more usable.

**Review by NewSoupVi — APPROVED** [review]
Seems good & necessary

**Pull request merged**